lib-js/test-case/util.ts

```diff
@@ -25,11 +25,17 @@ type FailureType =
   | "warning_difference"
   | "unnecessary_todo"
 
+/**
+ * The result of a test execution, along with metadata for failures and errors
+ */
 export interface TestResult {
-  type: "pass" | "fail" | "todo" | "skip"
+  type: "pass" | "fail" | "todo" | "skip" | "error"
+  // If `fail`, the type of failure, message, and possible diff
   failureType?: FailureType
   message?: string
   diff?: string
+  // If `error`, the thrown error
+  error?: Error
 }
```

Comment:
Consider making it explicit which fields go with which type:

```ts
export type TestResult =
  | ErrorResult
  | FailResult
  | { type: "pass" | "todo" | "skip" }

export interface ErrorResult {
  type: "error"
  error: Error
}

export interface FailResult {
  type: "fail"
  failureType: FailureType
  message: string
  diff: string
}
```

Reply:
I had it this way before, but it ended up being somewhat cumbersome (e.g. having to do an additional check in some cases even if you *know* a type is a `FailResult`). I can try and split it up again, however.

Reply:
okay so I recreated the issue. The problem is, in `interactor.ts`, we have a couple of functions that take `TestCase` as an argument. Methods on Interactor should only work on `FailResult`. So if we pass in the updated description, we get a bunch of errors like this:

```
lib-js/interactor.ts:74:42 - error TS2339: Property 'failureType' does not exist on type 'TestResult'.
  Property 'failureType' does not exist on type '{ type: "pass" | "todo" | "skip"; }'.
```

The problem is that all the methods in the Interactor take a `testCase` as an argument:

```
  {
    key: "G",
    description: (test: TestCase) => `Ignore test for ${test.impl} FOREVER`,
    requirement: (test: TestCase) => test.result().failureType !== "unnecessary_todo",
    resolve: (test: TestCase) => test.markIgnore(),
  },
```

There's no easy way to tell typescript that a particular `testCase` instance has a `FailResult`. So if you wanted it to work you'd have to do one of the following:

* Add checks in every method that `test.result()` is a `FailResult`, purely for type-checking (since we do a similar check in the functions that call these)
* Do an unsafe assertion that `test.result()` is a `FailResult`, (e.g. `(test.result() as FailResult).failureType`)
* Make `TestCase` take `TestResult` type as a generic parameter (which opens up another can of complexity)
* Pass the typed `FailResult` as an argument to all these submethods, which is extraneous and makes testing hard
* reorganize the structure of the interactor somehow

I do agree that splitting it up would be more type-safe, but given the complexity of the workarounds, would it be alright to just put in a TODO about the reasoning for now?

Reply:
Making it a TODO for now sounds fine.